Shell/InductionSchemeGenerator.cpp

```diff
@@ -0,0 +1,572 @@
+        }
+        i++;
+      }
+
+      auto isLit = t->isLiteral();
+      for (const auto& argTerms : argTermsList) {
+        Term* nt;
+        if (isLit) {
+          nt = Literal::create(static_cast<Literal*>(t), argTerms.begin());
+        } else {
+          nt = Term::create(t, argTerms.begin());
+        }
+        addScheme(lit, nt, templ, schemes);
+      }
+    } else {
+      addScheme(lit, t, templ, schemes);
+    }
+  } else {
+    for (unsigned i = 0; i < t->arity(); i++) {
+      actStack.push(active);
+    }
+  }
+}
+
+void StructuralInductionSchemeGenerator::generate(
+  const SLQueryResult& main,
+  const vset<pair<Literal*,Clause*>>& side,
+  vvector<pair<InductionScheme, OccurrenceMap>>& res)
+{
+  CALL("StructuralInductionSchemeGenerator()");
+
+  vvector<InductionScheme> schemes;
+  OccurrenceMap occMap;
+
+  Set<Term*> ta_terms;
+  SubtermIterator it(main.literal);
+  while (it.hasNext()) {
+    TermList ts = it.next();
+    ASS(ts.isTerm());
+    Term* t = ts.term();
+    unsigned f = t->functor();
+    if (Inferences::InductionHelper::isInductionTermFunctor(f) &&
+        Inferences::InductionHelper::isStructInductionFunctor(f)) {
+      ta_terms.insert(t);
+    }
+    occMap.add(main.literal, t, false);
+  }
+
+  Set<Term*>::Iterator taIt(ta_terms);
+  while (taIt.hasNext()) {
+    schemes.push_back(generateStructural(taIt.next()));
+  }
+
+  for (const auto& qr : side) {
+    SubtermIterator it(qr.first);
+    while (it.hasNext()) {
+      Term* t = it.next().term();
+      occMap.add(qr.first, t, false);
+    }
+  }
+
+  occMap.finalize();
+  for (const auto& sch : schemes) {
+    res.push_back(make_pair(sch, occMap.create_necessary(sch)));
+  }
+}
+
+InductionScheme StructuralInductionSchemeGenerator::generateStructural(Term* term)
+{
+  CALL("StructuralInductionSchemeGenerator::generateStructural");
+
+  TermAlgebra* ta = env.signature->getTermAlgebraOfSort(env.signature->getFunction(term->functor())->fnType()->result());
+  TermList ta_sort = ta->sort();
+  unsigned var = 1;
+  vmap<Term*, unsigned> inductionTerms;
+  inductionTerms.insert(make_pair(term, 0));
+  InductionScheme scheme(inductionTerms, true);
+
+  for (unsigned i = 0; i < ta->nConstructors(); i++) {
+    TermAlgebraConstructor* con = ta->constructor(i);
+    vvector<Substitution> recursiveCalls;
+    Substitution step;
+
+    unsigned arity = con->arity();
+    Stack<TermList> ta_vars;
+    Stack<TermList> argTerms;
+    for (unsigned i = 0; i < arity; i++) {
+      TermList x(var++, false);
+      argTerms.push(x);
+      if (con->argSort(i) == ta_sort) {
+        ta_vars.push(x);
+      }
+    }
+    Stack<TermList>::Iterator tvit(ta_vars);
+    while (tvit.hasNext()) {
+      recursiveCalls.emplace_back();
+      recursiveCalls.back().bind(0,tvit.next());
+    }
+    step.bind(0, TermList(Term::create(con->functor(),(unsigned)argTerms.size(), argTerms.begin())));
+    scheme.addCase(std::move(recursiveCalls), std::move(step));
+  }
```

Comment:
IIUC, while the scheme contains the 'term' (from arguments of generateStructural()) as an induction term, the cases only depend on the sort. However, the cases are always generated anew even if they were already generated for a different term of the same sort. Could this be optimized (eg. by caching the cases for each sort)?

Reply:
Good point. However, `InductionScheme`s are non-copyable because of the `DHMap` inside `Substitution`, so I ended up creating an intermediate container specific for these terms.

Reply:
Ok!